feat(route53resolver): added ResolverRuleAssociation

haarchri opened the pull request:
Signed-off-by: haarchri <chhaar30@googlemail.com>

<!--
Thank you for helping to improve Crossplane!

Please read through https://git.io/fj2m9 if this is your first time opening a
Crossplane pull request. Find us in https://slack.crossplane.io/messages/dev if
you need any help contributing.
-->

### Description of your changes

this feature fits for DNS management in a multi-account environment with Route 53 Resolver
- dns zone example.com lives in central dns aws-account
- subdomain zones like test.example.com lives in other aws-account(s) and needs resolver_rule_association to dns aws-account resolver rule (shared via ram)
- good documentation: https://aws.amazon.com/de/blogs/security/simplify-dns-management-in-a-multiaccount-environment-with-route-53-resolver/
- added route53_resolver_rule_association like terraform https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/route53_resolver_rule_association
- this resource is not generated with code-gen because we need the option to associate individual vpc with shared (ram) resolver rule

<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Crossplane issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

-->

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

`kubectl apply -f examples/route53resolver/associateresolverrule/associateresolverrule.yaml `


```
NAME                                                                                   READY   SYNCED   AGE
resolverruleassociation.route53resolver.aws.crossplane.io/sample-associateresolverrule   True    True     2m19s
```

```
kubectl describe resolverruleassociation.route53resolver.aws.crossplane.io/sample-associateresolverrule
Name:         sample-associateresolverrule
Namespace:    
Labels:       <none>
Annotations:  crossplane.io/external-create-pending: 2022-01-29T12:46:09+01:00
              crossplane.io/external-create-succeeded: 2022-01-29T12:46:09+01:00
              crossplane.io/external-name: rslvr-rrassoc-5666ff000b4044d58
API Version:  route53resolver.aws.crossplane.io/v1alpha1
Kind:         ResolverRuleAssociation
Metadata:
  Creation Timestamp:  2022-01-29T11:46:08Z
  Finalizers:
    finalizer.managedresource.crossplane.io
  Generation:  1
  Resource Version:  1483042
  UID:               2d1500c7-0612-4076-a10d-aea2507251cf
Spec:
  Deletion Policy:  Delete
  For Provider:
    Region:            eu-central-1
    Resolver Rule Id:  rslvr-rr-57432f045f54666aa
    Vpc Id:            vpc-33096b59
  Provider Config Ref:
    Name:  sandbox
Status:
  At Provider:
  Conditions:
    Last Transition Time:  2022-01-29T11:48:11Z
    Reason:                Available
    Status:                True
    Type:                  Ready
    Last Transition Time:  2022-01-29T11:46:09Z
    Reason:                ReconcileSuccess
    Status:                True
    Type:                  Synced
Events:
  Type    Reason                   Age    From                                                               Message
  ----    ------                   ----   ----                                                               -------
  Normal  CreatedExternalResource  2m51s  managed/resolverruleassociation.route53resolver.aws.crossplane.io  Successfully requested creation of external resource
```

<!--
Before reviewers can be confident in the correctness of this pull request, it
needs to tested and shown to be correct. Briefly describe the testing that has
already been done or which is planned for this change.
-->

[contribution process]: https://git.io/fj2m9

haarchri commented:
#679 we have one open PR adress this resource

muvaf commented on apis/route53resolver/manualv1alpha1/associateresolverrule_types.go:
```diff
@@ -0,0 +1,121 @@
+	// +crossplane:generate:reference:refFieldName=ResolverRuleIDRef
+	// +crossplane:generate:reference:selectorFieldName=ResolverRuleIDSelector
+	// +optional
+	ResolverRuleID *string `json:"resolverRuleID,omitempty"`
```
```suggestion
	ResolverRuleID *string `json:"resolverRuleId,omitempty"`
```

muvaf commented on apis/route53resolver/manualv1alpha1/associateresolverrule_types.go:
```diff
@@ -0,0 +1,121 @@
+	// the ResolverRuleID.
+	// +immutable
+	// +optional
+	ResolverRuleIDRef *xpv1.Reference `json:"resolverRuleIDRef,omitempty"`
```
```suggestion
	ResolverRuleIDRef *xpv1.Reference `json:"resolverRuleIdRef,omitempty"`
```

muvaf commented on apis/route53resolver/manualv1alpha1/associateresolverrule_types.go:
```diff
@@ -0,0 +1,121 @@
+	// to set the ResolverRuleID.
+	// +immutable
+	// +optional
+	ResolverRuleIDSelector *xpv1.Selector `json:"resolverRuleIDSelector,omitempty"`
```
```suggestion
	ResolverRuleIDSelector *xpv1.Selector `json:"resolverRuleIdSelector,omitempty"`
```

muvaf commented on examples/route53resolver/associateresolverrule/associateresolverrule.yaml:
```diff
@@ -0,0 +1,13 @@
+apiVersion: route53resolver.aws.crossplane.io/v1alpha1
+kind: AssociateResolverRule
```
```suggestion
kind: ResolverRuleAssociation
```
Could that be more inline with what exists?

haarchri replied:
i will set it to `RuleAssociation` then we have the same like provider-jet-aws

muvaf replied:
I'd lean towards how the AWS API names it in this repo, which is `ResolverRuleAssociation` I think looking at `GetResolverRuleAssociation` call.

muvaf commented on pkg/controller/route53resolver/associateresolverrule/hooks.go:
```diff
@@ -0,0 +1,162 @@
+		return errors.New(errUnexpectedObject)
+	}
+
+	cr.Status.SetConditions(xpv1.Deleting())
```
```suggestion
```

No need for this, since crossplane-runtime sets it now.

muvaf commented on pkg/controller/route53resolver/associateresolverrule/hooks.go:
```diff
@@ -0,0 +1,162 @@
+	return awsclient.Wrap(resource.Ignore(associateresolverrule.IsNotFound, err), errDelete)
+}
+
+func (e *external) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) {
+	_, ok := mg.(*manualv1alpha1.AssociateResolverRule)
+	if !ok {
+		return managed.ExternalUpdate{}, errors.New(errUnexpectedObject)
+	}
+
```
```suggestion
func (e *external) Update(_ context.Context, _ resource.Managed) (managed.ExternalUpdate, error) {
```

muvaf commented on pkg/controller/route53resolver/associateresolverrule/hooks.go:
```diff
@@ -0,0 +1,162 @@
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(manualv1alpha1.AssociateResolverRuleGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), newRoute53ResolverClientFn: associateresolverrule.NewRoute53ResolverClient}),
+			managed.WithInitializers(managed.NewDefaultProviderConfig(mgr.GetClient()), managed.NewNameAsExternalName(mgr.GetClient())),
```
```suggestion
			managed.WithInitializers(),
```
Its external name is computed by AWS.

muvaf commented on pkg/controller/route53resolver/associateresolverrule/hooks.go:
```diff
@@ -0,0 +1,162 @@
+
+	"github.com/crossplane/provider-aws/apis/route53resolver/manualv1alpha1"
+	"github.com/crossplane/provider-aws/pkg/clients/associateresolverrule"
+
```
```suggestion
```

muvaf commented on pkg/clients/associateresolverrule/associateresolverrule.go:
```diff
@@ -0,0 +1,76 @@
+	"context"
+	"errors"
+
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
```
Imports ordering.

muvaf reviewed (approved):
LGTM after resolving the comments!

haarchri commented:
fixed all suggestions, rebased & tested again

![image](https://user-images.githubusercontent.com/21083497/151660884-89ad5cc1-fa20-403c-a9ed-be490d91c29d.png)
